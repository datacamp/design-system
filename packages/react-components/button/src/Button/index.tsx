import * as Icons from '@datacamp/waffles-icons';
import { Text } from '@datacamp/waffles-text';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { computeDataAttributes } from '@datacamp/waffles-utils';
import { css, SerializedStyles } from '@emotion/core';
import { childrenOfType, nChildren } from 'airbnb-prop-types';
import PropTypes from 'prop-types';
import React, { ComponentProps, ReactElement, useRef, useState } from 'react';
import { useUIDSeed } from 'react-uid';
import { mergeRefs } from 'use-callback-ref';

import Tooltip from '../Tooltip';
import {
  baseColors,
  baseLoadingStyle,
  baseStyle,
  fontSizes,
  getAppearanceStyle,
  getDisabledStyle,
  getIconSize,
  getSize,
} from './buttonStyles';
import Spinner from './spinner';

type IconType = typeof Icons.AddCircleIcon; // Could use any Icon here
type IconElement = ReactElement<ComponentProps<IconType>, IconType>;

interface BaseButtonProps {
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
  /**
   * The button can be disabled by passing "disabled" as a prop.
   */
  disabled?: boolean;
  /**
   * By passing "loading" as a prop, the button will be disabled and it will
   * render a spinner.
   */
  loading?: boolean;
  /**
   * The size of the button to render.
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * When tooltipText is provided, this sets the colour of the tooltip. It
   * should be used make the tooltip visible on different colour backgrounds.
   */
  tooltipAppearance?: 'light' | 'dark';
  /**
   * When tooltipText is provided, this determines the position of that text.
   * "top", "bottom", "left", and "right" will position the tooltip at the
   * centre of that side. The other styles can be used to prevent the tooltip
   * from overflowing when it is longer than the button.
   */
  tooltipPosition?:
    | 'bottom'
    | 'top'
    | 'left'
    | 'right'
    | 'bottomRight'
    | 'bottomLeft'
    | 'topLeft'
    | 'topRight';
  /**
   * An additional description to show on hover or focus.
   */
  tooltipText?: string;
}

interface LinkButtonProps {
  href: string;
  target?: '_blank' | undefined;
  type: 'link';
}

interface ButtonButtonProps {
  onClick: () => void;
  type?: 'button' | undefined;
}

interface SubmitButtonProps {
  type: 'submit';
}

interface StringChildProps {
  ariaLabel?: string;
  children: string;
}

interface IconChildProps {
  ariaLabel: string;
  /**
   * When providing a react element as a child, it can only be one of the Icon
   * components exposed by @datacamp/waffles-icons
   */
  children: IconElement;
}

interface DefaultProps {
  appearance?: 'default';
  intent?: intents;
}

interface PrimaryProps {
  appearance: 'primary';
  intent?: intents | 'cta';
}

type intents = 'neutral' | 'danger' | 'success' | 'warning';

interface IconTextChildProps {
  ariaLabel?: string;
  children: [string, IconElement] | [IconElement, string];
}

export type ButtonProps = BaseButtonProps &
  (ButtonButtonProps | LinkButtonProps | SubmitButtonProps) &
  (StringChildProps | IconChildProps | IconTextChildProps) &
  (DefaultProps | PrimaryProps);

const InternalButton = (
  props: ButtonProps & { innerRef?: React.Ref<any> },
): React.ReactElement => {
  const {
    appearance = 'default',
    ariaLabel,
    children,
    className,
    dataAttributes,
    disabled = false,
    innerRef,
    intent = 'neutral',
    loading = false,
    size = 'medium',
    tooltipAppearance = 'dark',
    tooltipText,
    tooltipPosition = 'bottom',
  } = props;

  const buttonRef = useRef<HTMLElement>();
  const [hasFocus, setHasFocus] = useState(false);
  const [hasHover, setHasHover] = useState(false);
  const uidSeed = useUIDSeed();
  const tooltipId = uidSeed('button-tooltip');
  const outlineTextColor = tokens.color.opaque.greyDark.value.hex;
  const ctaTextColor = intent === 'cta' ? outlineTextColor : 'white';
  const outlineIconColor = disabled ? baseColors[intent] : 'currentColor';
  const textColor = appearance === 'primary' ? ctaTextColor : outlineTextColor;
  const iconColor = appearance === 'primary' ? ctaTextColor : outlineIconColor;

  const baseTextStyle = css(fontSizes[size], {
    color: loading ? 'transparent' : textColor,
    fontWeight: 'bold',
  });
  const margin =
    size === 'large' ? tokens.size.space[16].value : tokens.size.space[8].value;
  const getTextStyleWithMargin = (i: number | undefined): SerializedStyles => {
    if (i === 1) {
      return css(baseTextStyle, {
        marginLeft: margin,
      });
    }
    return css(baseTextStyle, {
      marginRight: margin,
    });
  };
  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  const buttonStyle = css(
    baseStyle,
    getAppearanceStyle(appearance, intent, !loading && !disabled),
    disabled && getDisabledStyle(appearance, intent),
    loading && baseLoadingStyle,
    React.Children.count(children) > 1 || typeof children === 'string'
      ? getSize(size)
      : getIconSize(size),
  );

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  const { Element, ...otherProps } = (() => {
    switch (props.type) {
      case 'link':
        return {
          Element: 'a',
          href: props.href,
          target: props.target,
        } as const;
      case 'submit':
        return { Element: 'button', type: 'submit' } as const;
      case 'button':
      default:
        return {
          Element: 'button',
          onClick: !loading ? props.onClick : undefined,
          type: 'button',
        } as const;
    }
  })();

  const tooltipVisible = !(disabled || loading) && (hasFocus || hasHover);

  return (
    <>
      <Element
        ref={innerRef ? mergeRefs([buttonRef, innerRef]) : buttonRef}
        aria-describedby={tooltipText && tooltipVisible ? tooltipId : undefined}
        aria-label={ariaLabel}
        className={className}
        css={buttonStyle}
        disabled={disabled || loading}
        onBlur={() => setHasFocus(false)}
        onFocus={() => setHasFocus(true)}
        onMouseEnter={() => setHasHover(true)}
        onMouseLeave={() => setHasHover(false)}
        {...otherProps}
        {...parsedDataAttributes}
      >
        {loading && (
          <Spinner
            css={{ position: 'absolute' }}
            inverted={appearance === 'primary' && intent !== 'cta'}
          />
        )}
        {React.Children.map(children, (child, i) =>
          typeof child === 'string' ? (
            <Text
              css={
                React.Children.count(children) > 1
                  ? getTextStyleWithMargin(i)
                  : baseTextStyle
              }
            >
              {child}
            </Text>
          ) : (
            React.cloneElement(child, {
              'aria-hidden': true, // hide icon from screen reader so only ariaLabel or button text is read.
              color: loading ? 'transparent' : child.props.color ?? iconColor,
              size: size === 'large' ? 24 : 18,
              title: '', // remove tooltip from icon within button
            })
          ),
        )}
      </Element>
      {tooltipText && (
        <Tooltip
          appearance={tooltipAppearance}
          id={tooltipId}
          position={tooltipPosition}
          target={buttonRef}
          visible={tooltipVisible}
        >
          {tooltipText}
        </Tooltip>
      )}
    </>
  );
};

// additional prop-types validation that the child is either icon or string
const iconValidator = childrenOfType(...Object.values(Icons));

InternalButton.propTypes = {
  /**
   * When the appearance is "primary", the button will have a filled colour.
   */
  appearance: PropTypes.oneOf(['default', 'primary']),
  /**
   * Set the aria-label on the rendered element. This is required when using a single icon as a child.
   */
  ariaLabel: PropTypes.string,
  /**
   * The content of the button. This can either be text, and icon, or a
   * combination of the two. When this is one of the components exposed by
   * @datacamp/waffles-icons, the button will render as a square. In this
   * situation ariaLabel is required.
   */
  children: PropTypes.oneOfType([
    nChildren(
      2,
      PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, iconValidator])),
    ),
    PropTypes.string,
    iconValidator,
  ]).isRequired,
  /**
   * The destination of the link. Only available when type="link".
   */
  href: PropTypes.string,
  /**
   * Defines the color of the button. The cta intent can be used only with
   * the primary appearance.
   */
  intent: PropTypes.oneOf(['neutral', 'danger', 'success', 'warning', 'cta']),
  /**
   * The callback fired when the button is clicked. Only available when
   * type="button".
   */
  onClick: PropTypes.func,
  /**
   * Specifies where to open the linked document. Only available when
   * type="link".
   */
  target: PropTypes.string,
  /**
   * Determines what is rendered, and what other props are available. The type
   * "submit" should be used for form submissions. The type "link" will render
   * an <a> tag, and should be used with an 'href'. And the type "button" should
   * be used with the 'onClick' prop.
   */
  type: PropTypes.oneOf(['button', 'link', 'submit']),
};

InternalButton.defaultProps = {
  appearance: 'default',
  disabled: false,
  intent: 'neutral',
  loading: false,
  size: 'medium',
  tooltipAppearance: 'dark',
  tooltipPosition: 'bottom',
  type: 'button',
};

const Button = React.forwardRef<any, ButtonProps>((props, ref) => (
  <InternalButton innerRef={ref} {...props} />
));

export default Button;
