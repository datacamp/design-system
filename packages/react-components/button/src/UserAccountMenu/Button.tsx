import { ChevronDownIcon, ChevronUpIcon } from '@datacamp/waffles-icons';
import tokens from '@datacamp/waffles-tokens';
import { css } from '@emotion/react';
import { useFocusRing } from '@react-aria/focus';
import { forwardRef } from 'react';

import AlertDot from './AlertDot';
import Avatar from './Avatar';
import { FULL_MENU_MEDIA_QUERY } from './constants';

const buttonStyle = css({
  alignItems: 'center',
  background: 'none',
  border: 0,
  borderRadius: 18,
  boxSizing: 'border-box',
  color: tokens.colors.navyText,
  cursor: 'pointer',
  display: 'inline-flex',
  [FULL_MENU_MEDIA_QUERY]: {
    ':hover': {
      backgroundColor: tokens.colors.beige200,
    },
    borderRadius: 15,
    height: 30,
    justifyContent: 'space-evenly',
    minWidth: 56,
    width: 56,
  },
  height: 36,
  minWidth: 36,
  outline: 0,
  padding: 0,
  width: 36,
});

const chevronStyle = css({
  display: 'none',
  [FULL_MENU_MEDIA_QUERY]: {
    display: 'block',
  },
});

type ButtonProps = {
  avatarUrl?: string;
  isOpen: boolean;
  showAlertDot?: boolean;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ avatarUrl, isOpen, showAlertDot = false, ...restProps }, ref) => {
    const { focusProps, isFocusVisible } = useFocusRing();

    return (
      <button
        css={css(
          buttonStyle,
          isFocusVisible && {
            backgroundColor: tokens.colors.beige200,
            boxShadow: `0 0 0 2px ${tokens.colors.blueDark}`,
          },
        )}
        {...restProps}
        {...focusProps}
        ref={ref}
      >
        <Avatar avatarUrl={avatarUrl} />
        {isOpen ? (
          <ChevronUpIcon aria-hidden={true} css={chevronStyle} />
        ) : (
          <ChevronDownIcon aria-hidden={true} css={chevronStyle} />
        )}
        {showAlertDot && <AlertDot />}
      </button>
    );
  },
);

export default Button;
