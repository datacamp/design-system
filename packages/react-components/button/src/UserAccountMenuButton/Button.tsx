import { ChevronDownIcon, ChevronUpIcon } from '@datacamp/waffles-icons';
import tokens from '@datacamp/waffles-tokens';
import { css } from '@emotion/react';
import { useFocusRing } from '@react-aria/focus';
import { ButtonHTMLAttributes, forwardRef } from 'react';

import Avatar from './Avatar';

const buttonStyle = css({
  '@media (min-width: 821px)': {
    ':hover': {
      backgroundColor: tokens.colors.beige200,
    },
    borderRadius: 15,
    height: 30,
    justifyContent: 'space-evenly',
    minWidth: 56,
    width: 56,
  },
  alignItems: 'center',
  background: 'none',
  border: 0,
  borderRadius: 18,
  color: tokens.colors.navyText,
  cursor: 'pointer',
  display: 'inline-flex',
  height: 36,
  minWidth: 36,
  outline: 0,
  padding: 0,
  width: 36,
});

const chevronStyle = css({
  '@media (min-width: 821px)': {
    display: 'block',
  },
  display: 'none',
});

type ButtonProps = {
  isOpen: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ isOpen, ...restProps }, ref) => {
    const { focusProps, isFocusVisible } = useFocusRing();

    return (
      <button
        css={css(
          buttonStyle,
          isFocusVisible && {
            backgroundColor: tokens.colors.beige200,
            boxShadow: `0 0 0 2px ${tokens.colors.blueDark}`,
          },
        )}
        {...restProps}
        {...focusProps}
        ref={ref}
      >
        <Avatar />
        {isOpen ? (
          <ChevronUpIcon aria-hidden={true} css={chevronStyle} />
        ) : (
          <ChevronDownIcon aria-hidden={true} css={chevronStyle} />
        )}
      </button>
    );
  },
);

export default Button;
