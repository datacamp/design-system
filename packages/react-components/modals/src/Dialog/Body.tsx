import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import React from 'react';

export interface BodyProps {
  children: React.ReactNode;
}

const Body = ({ children }: BodyProps): React.ReactElement => (
  <div
    css={{
      paddingBottom: tokens.size.space[32].value,
      paddingLeft: tokens.size.space[32].value,
      paddingRight: tokens.size.space[32].value,
    }}
  >
    {children}
  </div>
);

export default Body;
