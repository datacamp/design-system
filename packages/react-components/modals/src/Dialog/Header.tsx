import { Heading } from '@datacamp/waffles-text';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import React from 'react';

export interface HeaderProps {
  children: string;
}

const Header = ({ children }: HeaderProps): React.ReactElement => (
  <header
    css={{
      paddingBottom: tokens.size.space[24].value,
      paddingLeft: tokens.size.space[32].value,
      paddingRight: tokens.size.space[32].value,
      paddingTop: tokens.size.space[32].value,
    }}
  >
    <Heading as="h1" size={600}>
      {children}
    </Heading>
  </header>
);

export default Header;
