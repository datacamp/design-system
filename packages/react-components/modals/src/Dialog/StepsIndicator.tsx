import { ssrSafeNotFirstChildSelector } from '@datacamp/waffles-utils';
import { css } from '@emotion/react';
import _ from 'lodash';
import React from 'react';

const indicatorStyle = css({
  borderRadius: '50%',
  display: 'inline-block',
  height: 8,
  [ssrSafeNotFirstChildSelector]: {
    marginLeft: 8,
  },
  width: 8,
});

interface StepsIndicatorProps {
  currentStep?: number;
  totalSteps: number;
}

const StepsIndicator: React.FC<StepsIndicatorProps> = ({
  currentStep = 0,
  totalSteps,
}) => (
  <div
    aria-label={`Step ${currentStep} of ${totalSteps}`}
    css={{ lineHeight: '8px', marginTop: 16, padding: 0 }}
  >
    {_.times(totalSteps, (i) => (
      <div
        css={css(indicatorStyle, {
          backgroundColor:
            i >= currentStep ? 'rgba(255, 255, 255, 0.3)' : 'white',
        })}
        key={i}
      />
    ))}
  </div>
);

export default StepsIndicator;
