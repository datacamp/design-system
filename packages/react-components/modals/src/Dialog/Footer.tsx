import Button from '@datacamp/waffles-button';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { childrenOfType } from 'airbnb-prop-types';
import React from 'react';

export interface FooterProps {
  children: React.ReactElement;
}

const Footer = ({ children }: FooterProps): React.ReactElement => (
  <div
    css={{
      boxShadow: '0px -1px 0px 0px rgb(230, 234, 235)',
      paddingBottom: tokens.size.space[16].value,
      paddingTop: tokens.size.space[16].value,
      textAlign: 'center',
    }}
  >
    {children}
  </div>
);

Footer.propTypes = {
  children: childrenOfType(Button).isRequired,
};

export default Footer;
