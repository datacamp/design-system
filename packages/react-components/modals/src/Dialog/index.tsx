import React from 'react';

import BaseDialog, { CloseOrigin } from '../BaseDialog';
import Header from './Header';

interface DialogProps {
  /**
   * The content of the dialog to render
   */
  children: React.ReactNode;
  /**
   * When true, the close button in the top right corner will not be shown
   */
  hideCloseButton?: boolean;
  /**
   * Whether to hide or show the Dialog
   */
  isOpen: boolean;
  /**
   * Called when the modal is closed
   * @param origin The original source of the request to close.
   */
  onClose: (origin: CloseOrigin) => void;
  /**
   * When true (default), the Dialog will close when the Esc key is pressed.
   */
  shouldCloseOnEsc?: boolean;
  /**
   * When true (default), the Dialog will close when the bckground overlay is
   * clicked.
   */
  shouldCloseOnOverlayClick?: boolean;
  /**
   * Forces a specific width
   */
  width?: string;
}

const Dialog = ({
  children,
  hideCloseButton = false,
  isOpen,
  onClose,
  shouldCloseOnEsc = true,
  shouldCloseOnOverlayClick = true,
  width,
}: DialogProps): React.ReactElement => (
  <BaseDialog
    contentLabel="Dialog"
    hideCloseButton={hideCloseButton}
    isOpen={isOpen}
    onClose={onClose}
    shouldCloseOnEsc={shouldCloseOnEsc}
    shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
    width={width}
  >
    {children}
  </BaseDialog>
);

Dialog.Header = Header;

export default Dialog;
