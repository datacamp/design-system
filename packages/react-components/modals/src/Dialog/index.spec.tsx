import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import Button, { ButtonGroup } from '@datacamp/waffles-button';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { fireEvent, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';

import { setAppElement } from '..';
import Dialog from '.';

describe('<Dialog />', () => {
  const baseContainer = document.createElement('div');
  setAppElement(baseContainer);
  const getRenderOptions = (): RenderOptions => ({
    baseElement: document.body,
    container: document.body.appendChild(baseContainer),
  });

  describe('isOpen', () => {
    it('renders the modal via a portal when true', async () => {
      const { container, getByText, baseElement, getByRole } = await axeRender(
        <Dialog onClose={() => {}} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      expect(container).toHaveAttribute('aria-hidden', 'true'); // app gets hidden when open

      const modalElement = getByRole('dialog');
      expect(modalElement).toContainElement(
        getByText('children') as HTMLElement
      );
      expect(baseElement).toMatchSnapshot();
    });

    it('mounts nothing when false', async () => {
      const { container, baseElement, queryByRole } = await axeRender(
        <Dialog isOpen={false} onClose={() => {}}>
          children
        </Dialog>,
        getRenderOptions()
      );

      expect(container).not.toHaveAttribute('aria-hidden', 'true'); // app is interactive when hidden
      expect(queryByRole('dialog')).not.toBeInTheDocument();
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('close button', () => {
    it('closes when clicking the visible X button by default', async () => {
      const onClose = jest.fn();

      const { getByTitle } = await axeRender(
        <Dialog onClose={onClose} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      const closeButton = getByTitle('Cross') as HTMLElement;

      userEvent.click(closeButton);
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('closeButton');
    });

    it('closes when clicking the visible X button when hideCloseButton=false', async () => {
      const onClose = jest.fn();

      const { getByTitle } = await axeRender(
        <Dialog hideCloseButton={false} onClose={onClose} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      const closeButton = getByTitle('Cross') as HTMLElement;

      userEvent.click(closeButton);
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('closeButton');
    });

    it("doesn't show the X button when hideCloseButton is true", async () => {
      const onClose = jest.fn();

      const { queryByTitle } = await axeRender(
        <Dialog onClose={onClose} hideCloseButton isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      expect(queryByTitle('Cross')).not.toBeInTheDocument();
    });
  });

  describe('background overlay', () => {
    it('closes when clicking the background overlay by default', async () => {
      const onClose = jest.fn();
      const { baseElement } = await axeRender(
        <Dialog onClose={onClose} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      // need to use querySelector since there is no real content/label for the background
      const backgroundOverlayElement = baseElement.querySelector(
        '.modal-overlay'
      ) as HTMLElement;

      userEvent.click(backgroundOverlayElement);
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('overlayClick');
    });

    it('closes when clicking the background overlay when shouldCloseOnOverlayClick=true', async () => {
      const onClose = jest.fn();
      const { baseElement } = await axeRender(
        <Dialog onClose={onClose} isOpen shouldCloseOnOverlayClick>
          children
        </Dialog>,
        getRenderOptions()
      );

      // need to use querySelector since there is no real content/label for the background
      const backgroundOverlayElement = baseElement.querySelector(
        '.modal-overlay'
      ) as HTMLElement;

      userEvent.click(backgroundOverlayElement);
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('overlayClick');
    });

    it("doesn't when clicking the background overlay when shouldCloseOnOverlayClick=false", async () => {
      const onClose = jest.fn();
      const { baseElement } = await axeRender(
        <Dialog onClose={onClose} shouldCloseOnOverlayClick={false} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      // need to use querySelector since there is no real content/label for the background
      const backgroundOverlayElement = baseElement.querySelector(
        '.modal-overlay'
      ) as HTMLElement;

      userEvent.click(backgroundOverlayElement);
      expect(onClose).not.toHaveBeenCalled();
    });
  });

  describe('escape key', () => {
    it('closes when clicking escape by default', async () => {
      const onClose = jest.fn();
      const { getByRole } = await axeRender(
        <Dialog onClose={onClose} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      const dialogElement = getByRole('dialog') as HTMLElement;

      fireEvent.keyDown(dialogElement, {
        key: 'ESC',
        keyCode: 27,
      });

      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('escKey');
    });

    it('closes when clicking escape if shouldCloseOnEsc=true', async () => {
      const onClose = jest.fn();
      const { getByRole } = await axeRender(
        <Dialog onClose={onClose} isOpen shouldCloseOnEsc>
          children
        </Dialog>,
        getRenderOptions()
      );

      const dialogElement = getByRole('dialog') as HTMLElement;

      fireEvent.keyDown(dialogElement, {
        key: 'ESC',
        keyCode: 27,
      });

      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('escKey');
    });

    it("doesn't close when clicking escape if shouldCloseOnEsc=false", async () => {
      const onClose = jest.fn();
      const { getByRole } = await axeRender(
        <Dialog onClose={onClose} shouldCloseOnEsc={false} isOpen>
          children
        </Dialog>,
        getRenderOptions()
      );

      const dialogElement = getByRole('dialog') as HTMLElement;

      fireEvent.keyDown(dialogElement, {
        key: 'ESC',
        keyCode: 27,
      });

      expect(onClose).not.toHaveBeenCalled();
    });
  });

  it('sets the width', async () => {
    const onClose = jest.fn();
    const width = '200px';
    const { getByRole } = await axeRender(
      <Dialog onClose={onClose} width={width} isOpen>
        children
      </Dialog>,
      getRenderOptions()
    );

    const dialogElement = getByRole('dialog') as HTMLElement;
    expect(dialogElement).toHaveStyle(`width: ${width};`);
  });

  it('renders with a Header', async () => {
    const testTitle = 'test title';

    const { getByText } = await axeRender(
      <Dialog onClose={() => {}} isOpen>
        <Dialog.Header>{testTitle}</Dialog.Header>
        children
      </Dialog>,
      getRenderOptions()
    );

    expect(getByText(testTitle)).toBeInTheDocument();

    // expect(baseElement).toMatchSnapshot();
  });

  it('renders with a Header, Body & Footer', async () => {
    const testTitle = 'test title';

    const { getByText } = await axeRender(
      <Dialog onClose={() => {}} isOpen>
        <Dialog.Header>{testTitle}</Dialog.Header>
        <Dialog.Body>children</Dialog.Body>
        <Dialog.Footer>
          <Button onClick={() => {}}>Button</Button>
        </Dialog.Footer>
      </Dialog>,
      getRenderOptions()
    );

    expect(getByText(testTitle)).toBeInTheDocument();
    expect(getByText('children')).toBeInTheDocument();
    expect(getByText('Button')).toBeInTheDocument();

    // expect(baseElement).toMatchSnapshot();
  });

  it('renders two buttons in the Footer with space in between', async () => {
    const { container } = await axeRender(
      <Dialog.Footer>
        <Button onClick={() => {}}>Button1</Button>
        <Button onClick={() => {}}>Button2</Button>
      </Dialog.Footer>,
      getRenderOptions()
    );

    expect(container.firstChild).toHaveStyle(`justify-content: space-between`);
    expect(container.firstChild).toHaveStyle(
      `paddingLeft: ${tokens.size.space[16].value}`
    );
    expect(container.firstChild).toHaveStyle(
      `paddingRight: ${tokens.size.space[16].value}`
    );

    // expect(baseElement).toMatchSnapshot();
  });

  it('renders a ButtonGroup in the Footer', async () => {
    const { getByText } = await axeRender(
      <Dialog.Footer>
        <ButtonGroup>
          <Button onClick={() => {}}>Button1</Button>
          <Button onClick={() => {}}>Button2</Button>
        </ButtonGroup>
      </Dialog.Footer>,
      getRenderOptions()
    );

    expect(getByText('Button1')).toBeInTheDocument();
    expect(getByText('Button2')).toBeInTheDocument();
    expect(getByText('Button2')).toHaveStyle(
      `marginLeft: ${tokens.size.space[16].value}`
    );

    // expect(baseElement).toMatchSnapshot();
  });
});
