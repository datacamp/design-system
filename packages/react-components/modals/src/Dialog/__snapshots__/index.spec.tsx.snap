// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> isOpen mounts nothing when false 1`] = `
<body
  class=""
>
  <div
    class="ReactModalPortal"
  />
  <div />
  <div
    class="ReactModalPortal"
  />
</body>
`;

exports[`<Dialog /> isOpen renders the modal via a portal when true 1`] = `
.emotion-0 {
  overflow: hidden;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: rgba(5, 25, 45, 0.8);
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  left: 0;
  opacity: 0;
  position: fixed;
  right: 0;
  top: 0;
  -webkit-transition: opacity 0.6s cubic-bezier(0.19, 1, 0.22, 1);
  transition: opacity 0.6s cubic-bezier(0.19, 1, 0.22, 1);
  z-index: 900;
}

.emotion-2 {
  opacity: 1!important;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  min-width: 0;
  opacity: 0;
  outline: none;
  position: relative;
  -webkit-transform: scale(0.5);
  -moz-transform: scale(0.5);
  -ms-transform: scale(0.5);
  transform: scale(0.5);
  -webkit-transition: 0.4s cubic-bezier(0.19, 1, 0.22, 1);
  transition: 0.4s cubic-bezier(0.19, 1, 0.22, 1);
  box-sizing: border-box;
  max-height: 100%;
  padding: 8px;
  width: 616px;
}

.emotion-4 {
  opacity: 1!important;
  -webkit-transform: none!important;
  -moz-transform: none!important;
  -ms-transform: none!important;
  transform: none!important;
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
  border: none;
  border-radius: 2px;
  color: rgb(5, 25, 45);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 32px;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline: none;
  padding: 0;
  width: 32px;
  z-index: 1;
  position: absolute;
  right: 16px;
  top: 16px;
}

.emotion-5:active {
  color: rgb(3, 239, 98);
}

.emotion-5:disabled {
  color: #D1D3D8;
}

.emotion-5:disabled:hover,
.emotion-5:disabled:focus {
  background-color: transparent;
}

.emotion-5:focus {
  background-color: rgb(239, 239, 239);
}

.emotion-5:hover {
  background-color: rgb(239, 239, 239);
}

.emotion-6 {
  background-color: #ffffff;
  border-radius: 4px;
  -webkit-transition: all 600ms cubic-bezier(0.075, 0.82, 0.165, 1);
  transition: all 600ms cubic-bezier(0.075, 0.82, 0.165, 1);
  box-shadow: 0 0 2px 0 rgba(5, 25, 45, 0.25),0 8px 12px -4px rgba(5,25,45,0.3);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-height: 100%;
  overflow: hidden;
  width: 100%;
}

<body
  class="emotion-0"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
  />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="modal-overlay emotion-1 emotion-2"
    >
      <div
        aria-label="Dialog"
        aria-modal="true"
        class="emotion-3 emotion-4"
        role="dialog"
        tabindex="-1"
      >
        <button
          aria-label="Close"
          class="emotion-5"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="18"
            role="img"
            viewBox="0 0 18 18"
            width="18"
          >
            <title>
              Cross
            </title>
            <path
              d="M9.005 7.625l4.83-4.83a.976.976 0 011.38 1.38l-4.83 4.83 4.82 4.82a.976.976 0 11-1.38 1.38l-4.82-4.82-4.83 4.83a.976.976 0 01-1.38-1.38l4.83-4.83-4.84-4.84a.976.976 0 111.38-1.38l4.84 4.84z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <div
          class="emotion-6"
        >
          children
        </div>
      </div>
    </div>
  </div>
</body>
`;
